unittest/server/src/ivi_controller_uinttests.cpp

```diff
@@ -0,0 +1,3475 @@
+	"wl_buffer", 1,
+	0, NULL,
+	0, NULL,
+};
+
+WL_EXPORT const struct wl_interface wl_output_interface = {
+	"wl_output", 4,
+	0, NULL,
+	0, NULL,
+};
+
+struct wl_resource * custom_wl_resource_from_link(struct wl_list *link)
+{
+	struct wl_resource *resource;
+
+	return wl_container_of(link, resource, link);
+}
+
+struct wl_list * custom_wl_resource_get_link(struct wl_resource *resource)
+{
+	return &resource->link;
+}
+
+#include "ivi-controller.c"
+}
+
+static constexpr uint8_t MAX_NUMBER = 2;
+static uint32_t g_SurfaceCreatedCount = 0;
+static uint32_t custom_get_id_of_surface(struct ivi_layout_surface *ivisurf)
+{
+    return ivisurf->id_surface;
+}
+
+static uint32_t custom_get_id_of_layer(struct ivi_layout_layer *ivilayer)
+{
+    return ivilayer->id_layer;
+}
+
+static int32_t custom_get_layers_on_screen(struct weston_output *output, int32_t *pLength,
+					                       struct ivi_layout_layer ***ppArray)
+{
+    struct ivi_layout_layer ivilayer;
+    ivilayer.id_layer = 1;
+
+    *pLength = 1;
+    *ppArray = (ivi_layout_layer**)malloc(sizeof(struct ivi_layout_layer *));
+    (*ppArray)[0] = &ivilayer;
+
+    return 0;
+}
+
+static int32_t custom_get_surfaces_on_layer(struct ivi_layout_layer *ivilayer, int32_t *pLength,
+				                            struct ivi_layout_surface ***ppArray)
+{
+    struct ivi_layout_surface ivisurface;
+    ivisurface.id_surface = 1;
+
+    *pLength = 1;
+    *ppArray = (ivi_layout_surface**)malloc(sizeof(struct ivi_layout_surface *));
+    (*ppArray)[0] = &ivisurface;
+
+    return 0;
+}
+
+static int32_t custom_surface_get_size(struct ivi_layout_surface *ivisurf, int32_t *width, int32_t *height,
+			                           int32_t *stride)
+{
+    *width = 1;
+    *height = 1;
+    *stride = 1;
+
+    return 0;
+}
+
+static int custom_read_pixels(struct weston_output *output,
+			                  pixman_format_code_t format, void *pixels,
+			                  uint32_t x, uint32_t y,
+			                  uint32_t width, uint32_t height)
+{
+    return x == 0 || y == 0;
+}
+
+static void surface_create_callback(struct wl_listener *listener, void *data)
+{
+    g_SurfaceCreatedCount++;
+}
+
+class ControllerTests : public ::testing::Test
+{
+public:
+    void SetUp()
+    {
+        IVI_LAYOUT_FAKE_LIST(RESET_FAKE);
+        SERVER_API_FAKE_LIST(RESET_FAKE);
+        get_id_of_surface_fake.custom_fake = custom_get_id_of_surface;
+        get_id_of_layer_fake.custom_fake = custom_get_id_of_layer;
+        get_layers_on_screen_fake.custom_fake = custom_get_layers_on_screen;
+        get_surfaces_on_layer_fake.custom_fake = custom_get_surfaces_on_layer;
+        surface_get_size_fake.custom_fake = custom_surface_get_size;
+        g_SurfaceCreatedCount = 0;
+        init_controller_content();
+    }
+
+    void TearDown()
+    {
+        deinit_controller_content();
+    }
+
+    void init_controller_content()
+    {
+        uint8_t l_fakeResource = 0, l_fakeClient = 0;
```

Comment:
refactor this function to more simple:
- Fix scene graph
- consider to add new internal functions for: add/remove/get_surface, add/remove/get_layer, add/remove/get_screen, add/remove/get_surface_properties, add/remove/get_surface_notification, add/remove/get_layer_properties, add/remove/get_layer_notfication ... TOO MUCH :(

Reply:
we will consider to do this thing in refactor action in the future